lib/elixir/src/elixir_erl_for.erl

```diff
@@ -131,28 +146,47 @@ build_into(Ann, Clauses, Expr, Into, Fun, Acc, S) ->
         [{tuple, Ann, [Kind, Reason, {var, Ann, '_'}]}],
         [],
         [{match, Ann, Stack, elixir_erl:remote(Ann, erlang, get_stacktrace, [])},
-         {call, Ann, Fun, [Acc, {atom, Ann, halt}]},
+         {call, Ann, Fun, [retrieve_acc(Ann, Acc, Uniq), {atom, Ann, halt}]},
          elixir_erl:remote(Ann, erlang, raise, [Kind, Reason, Stack])]}],
       []},
 
   {{block, Ann, [MatchExpr, TryExpr]}, SD}.
 
 %% Helpers
 
-build_reduce(Clauses, Expr, false, Acc, S) ->
-  build_reduce_clause(Clauses, Expr, {nil, 0}, Acc, S);
-build_reduce(Clauses, Expr, {nil, Ann} = Into, Acc, S) ->
-  ListExpr = {cons, Ann, Expr, Acc},
-  elixir_erl:remote(Ann, lists, reverse,
-    [build_reduce_clause(Clauses, ListExpr, Into, Acc, S)]);
-build_reduce(Clauses, Expr, {bin, Ann, _} = Into, Acc, S) ->
-  BinExpr = {cons, Ann, Acc, Expr},
-  elixir_erl:remote(Ann, erlang, iolist_to_binary,
-    [build_reduce_clause(Clauses, BinExpr, Into, Acc, S)]).
-
-build_reduce_clause([{enum, Meta, Left, Right, Filters} | T], Expr, Arg, Acc, S) ->
+retrieve_acc(_Ann, Acc, false) ->
+  Acc;
+retrieve_acc(Ann, Acc, true) ->
+  elixir_erl:remote(Ann, erlang, element, [{integer, Ann, 1}, Acc]).
+
+build_reduce(_Ann, Clauses, InnerFun, Expr, Into, false, Acc, S) ->
+  InnerExpr = InnerFun(Expr, Acc),
+  build_reduce_each(Clauses, InnerExpr, Into, Acc, S);
+build_reduce(Ann, Clauses, InnerFun, Expr, Into, true, Acc, S) ->
+  {Value, SV} = build_var(Ann, S),
+  {IntoAcc, SI} = build_var(Ann, SV),
+  {UniqAcc, SU} = build_var(Ann, SI),
```

Comment:
We always need to return the scope when we modify it, otherwise we get variable conflicts.